test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,1493 @@
+			curTestPatternName := testPatternName
+			curTestPattern := testPattern
+
+			for driverName, initDriver := range testDrivers {
+				curDriverName := driverName
+				curInitDriver := initDriver
+				Context(fmt.Sprintf("Driver: %s, Test: %s, TestPattern: %s",
+					curDriverName, curTestCaseName, curTestPatternName), func() {
+					var (
+						driver       testDriver
+						testCase     testCase
+						testResource testResource
+						testInput    testInput
+					)
+
+					BeforeEach(func() {
+						driver = curInitDriver(f, config, node)
+						// Skip tests before driver creates test resource to avoid unnecessary initialization
+						skipTestForDriver(curTestPattern, driver)
+
+						driver.createDriver()
+						testCase = curTestCase
+						testResource = testCase.initTestResource()
+						testResource.setup(driver, curTestPattern)
+						testInput = testCase.createTestInput(curTestPattern, testResource)
+					})
+
+					AfterEach(func() {
+						testResource.cleanup()
+						driver.cleanupDriver()
+					})
+
+					curTestFunc(&testInput)
+				})
+			}
+		}
+	}
+})
+
+type testDriver interface {
+	createDriver()
+	cleanupDriver()
+	getDriverInfo() driverInfo
+}
+
+type inlineVolumeTestDriver interface {
+	testDriver
+	getVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
+}
+
+type staticPVTestDriver interface {
+	testDriver
+	getPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource
+}
+
+type dynamicPVTestDriver interface {
+	testDriver
+	getDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
+}
+
+type driverInfo struct {
+	name                      string
+	fsGroup                   *int64
+	podSec                    v1.PodSecurityContext
+	privilegedSecurityContext bool
+	maxFileSize               int64
+
+	// supportablity for driver
+	supportedFsType  map[string]bool
+	isBlockSupported bool
+
+	// volumes test specific parameters
+	expectedContent string
+	needInjectHTML  bool
+
+	// volumeIO test specific parameter
+	testFile string
+
+	f      *framework.Framework
+	config framework.VolumeTestConfig
+	node   v1.Node
+}
+
+type testCase interface {
+	getTestPatterns() map[string]testPattern
+	initTestResource() testResource
+	createTestInput(testPattern, testResource) testInput
+	getTestFunc() func(*testInput)
+}
+
+type testVolType string
+
+var (
+	inlineVolume testVolType = "inlineVolume"
+	staticPV     testVolType = "staticPV"
+	dynamicPV    testVolType = "dynamicPV"
+)
+
+type testPattern struct {
+	testVolType testVolType
+	fsType      string
+	volBindMode storagev1.VolumeBindingMode
+	volMode     v1.PersistentVolumeMode
+}
+
+type testResource interface {
+	setup(testDriver, testPattern)
+	cleanup()
+}
+
+type testInput interface {
+	isTestInput() bool
+}
+
+func skipTestForDriver(testPattern testPattern, driver testDriver) {
+	driverInfo := driver.getDriverInfo()
+	// Check if testVolType is supported by driver
+	if testPattern.testVolType == inlineVolume {
```

Comment:
might be better to do a switch statement, and error on the default case

Reply:
Thank you for your advice. I will fix it.